feat(rpc): network verification via commitment header

Mirko-von-Leipzig opened the pull request:
This PR adds support for verifying the network identity for RPC clients.

This PR also changes the existing format for the version checking:

```
# Original
application/vnd.miden+grpc.x.y.z

# New
application/vnd.miden; version=x.y.z; genesis=0xABC
```

The now obsolete format gets rejected now which I think is okay. The new format still allows for an optional `+grpc` suffix.

The original version checking behavior is retained, though we may want to consider switching to `Caret` instead of `Exact` but that's separate from this addition.

This PR also improves the correctness of the accept header verification by:

- accepting wildcards `*/*` and `application/*` media-types
- rejecting zero-weighted media-types `q=0`

~Similar to the RPC version check added in #844, we now check the network the client thinks its on. The client can now set the genesis block's commitment in the `miden-network` header, and the RPC service will reject the request if there is a mismatch.~

~This does add an additional header, and maybe instead we should add it to the `accept` header? I'm just unsure what this format should look like; and we probably want to do additional parsing instead.~

Closes #1066.

sergerad commented on crates/rpc/src/server/accept.rs:
```diff
@@ -162,61 +193,88 @@ impl<'a> TryFrom<&'a str> for AcceptHeaderValue<'a> {
-        version.patch += 1;
-        assert!(a.version_req.matches(&version));
+    impl HeaderVerificationLayer {
+        fn for_tests() -> Self {
```
can we `#[cfg(test)]` this fn?

Mirko-von-Leipzig replied:
It already is because its in `mod tests`

sergerad commented on crates/rpc/src/server/accept.rs:
```diff
@@ -162,61 +193,88 @@ impl<'a> TryFrom<&'a str> for AcceptHeaderValue<'a> {
-        };
-        assert_eq!(AcceptHeaderValue::try_from(input), Ok(expected));
+        let uut = HeaderVerificationLayer::for_tests();
+        uut.verify(&headers).unwrap_err();
```
Could we test the exact error variants coming out? Given you have created the error enums now.

sergerad replied:
> For example, in the case where both mismatch, which error should it be? Could be either..

It would be well defined which error would be returned if both mismatched. Checking the error proves our logic is working precisely as expected. We could erroneously return invalid semver variant for an input with a correct semver for example. Unless I'm missing something.

sergerad replied:
I had a crack at adding the expected error variants into the cases but I think its blocked on variants like 
```rust
    VersionParsing(#[source] semver::Error, String),
```

Because `semver::Error` and its constructors etc are not public so we can't recreate them for the tests.

sergerad commented on crates/rpc/src/server/api.rs:
```diff
@@ -88,6 +91,54 @@ impl RpcService {
+    /// Automatically retries until the store connection becomes available.
+    pub async fn get_genesis_header_with_retry(&self) -> anyhow::Result<BlockHeader> {
+        let mut retry_counter = 0;
+        loop {
```
Looks like we are now doing this pattern 3x in the code. Wondering if we could implement a utility fn that runs a closure with retries.

Mirko-von-Leipzig replied:
#1072 contains some of my thoughts on this; but yeah its something I'm hoping will fall out of all the gRPC refactoring.

sergerad commented on crates/rpc/src/server/api.rs:
```diff
@@ -88,6 +91,54 @@ impl RpcService {
+                        "connection failed while subscribing to the mempool, retrying"
+                    );
+
+                    retry_counter += 1;
```
Does this loop forever until it succeeds? Should we add a limit and error out instead? This would be something handled by the utility fn I mentioned  [here](https://github.com/0xMiden/miden-node/pull/1084/files#r2205996865).

sergerad replied:
One more thought on this - with the loop + WARN approach we would need individual alerts for each such situation. Whereas crash loops would be alerted on by my general alerts that are not so implementation-aware.

sergerad commented on crates/rpc/src/server/accept.rs:
```diff
@@ -1,61 +1,146 @@
+/// The former is taken from the `accept` header with the expected format
+/// `application/vnd.miden.<major>.<minor>.<patch>+grpc`.
+///
+/// The latter expects the network's genesis block's commitment in the `miden-network` header.
```
Wondering if we should just re-use the accept header for this. If we plan on putting more things like this in place, I don't think we should. But if this is the only other verification for client/server/network matching, then we could integrate it e.g.
```
application/vnd.miden.<network>.<major>.<minor>.<patch>+grpc

sergerad replied:
> So having two separate "media type" as part of the header would be incorrect afaict
Yep.

> q is intended to let the user set priorities, a higher value means they prefer this content type.

Is this to say that `; network=<..>` would be inappropriate?

bobbinth replied:
Is `network` self-explanatory enough here? maybe it should be `genesis=0x123...`? Though, probably doesn't matter too much.

sergerad replied:
Yea network is a bit misleading. Genesis is better.

sergerad replied:
Yes I think that is practical for us and also idiomatic as far as I could tell from the relevant RFCs.

bobbinth reviewed (commented):
Thank you! Looks good! Not a review but I did leave a couple of questions/comments inline.

> This does add an additional header, and maybe instead we should add it to the `accept` header? I'm just unsure what this format should look like; and we probably want to do additional parsing instead.

I haven't thought about it too much, but it does feel like maybe handling everything through the `accept` header is a bit more "natural" - i.e., we have a single header which tells us if the request should be accepted or rejected.

On the other hand, this probably complicates parsing a bit and makes it more difficult to handle partial scenarios (e.g., missing version info but present genesis commitment  or vise-versa). But it is also not clear if we should care about partial scenarios - and in that case, maybe channeling everything through the `accept` header is a better option.

Mirko-von-Leipzig commented on crates/rpc/src/server/mod.rs:
```diff
@@ -55,10 +61,14 @@ impl Rpc {
 
         info!(target: COMPONENT, endpoint=?self.listener, store=%self.store, block_producer=?self.block_producer, "Server initialized");
 
+        let rpc_version = env!("CARGO_PKG_VERSION");
+        let rpc_version =
+            semver::Version::parse(rpc_version).context("failed to parse crate version")?;
```
> We also use the package version for the status endpoint. Should we centralize this somehow? For example, should we put the version into the RpcService struct?

I don't think this is complicated enough to really warrant de-duping it. And in this case its required in a layer and not as part of the service itself so this wouldn't really change much.

> A somewhat related thought: should we also include the genesis (or genesis_commitment) field into the response from the status endpoint?

We can; would be less fluff than fetching the genesis block's header specifically which would be nice for manual `grpcurl` style checks. #1088

bobbinth replied:
> I don't think this is complicated enough to really warrant de-duping it. And in this case its required in a layer and not as part of the service itself so this wouldn't really change much.

My thinking here was that maybe in the future the version we'd want to use is not the same as the version of the crate (e.g., we could have "protocol versions" which are different from crate versions) and so it would be nice to have a central place where it is defined. But also, it is probably something we can do if/when we actually use something other than the crate version.

tomyrd commented on crates/rpc/src/server/accept.rs:
```diff
@@ -75,68 +160,14 @@ where
-        let accept = match AcceptHeaderValue::try_from(accept_str) {
-            Ok(value) => value,
-            Err(e) => {
-                // If the accept header value is different from expected, don't enforce the version
-                // requirement.
-                debug!(target: COMPONENT, "Failed to parse accept header value: {}", e);
-                return self.inner.call(request).boxed();
```
In the old version the accept header was not rejected when it failed to parse. This is not the same in the new version and caused a specific issue with the web client.
Apparently the web client sends an initial request with a `*/*` accept header which is causing the error. [We noticed this ](https://github.com/0xMiden/miden-client/pull/922#issuecomment-2910454361) when we originally added the accept header. Could we add this behavior to the new version? Or maybe add a special case for `*/*`.

drahnr commented on crates/rpc/src/server/accept.rs:
```diff
@@ -1,61 +1,168 @@
+/// As part of the accept header's standard, all media ranges are examined in quality weighting
+/// order until a matching content type is found. This means that the client can set multiple
+/// `application/vnd.miden` values and each will be tested until one passes. If none pass, the
+/// request is rejected.
```
Q: Should we limit the number?

drahnr commented on crates/rpc/src/tests.rs:
```diff
@@ -77,14 +81,7 @@ async fn rpc_server_rejects_requests_with_accept_header_invalid_version() {
-                .message()
-                .contains("Client / server version mismatch"),
-        );
+        assert!(response.as_ref().err().unwrap().message().contains("server does not support"),);
```
nit: I find the output of `assert_matches::assert_matches!` much easier to interpret any issues

igamigo commented on crates/rpc/src/server/accept.rs:
```diff
@@ -1,61 +1,168 @@
-        let version = Version::parse(version)?;
+#[derive(Debug, thiserror::Error)]
+enum AcceptHeaderError {
+    #[error("Header value could not be parsed as a UTF8 string")]
```
```suggestion
    #[error("header value could not be parsed as a UTF8 string")]
```

igamigo commented on crates/rpc/src/server/accept.rs:
```diff
@@ -75,148 +182,232 @@ where
+    Type(&'a str),
+}
+
+#[derive(Debug, Eq, Hash)]
```
nit: `CaselessKey("q")` and `CaselessKey("Q")` would hash to different things. Maybe you were keeping this in mind but just noting it just in case.

igamigo commented on crates/rpc/src/server/accept.rs:
```diff
@@ -75,148 +182,232 @@ where
+        match self {
+            MediaType::Wildcard => false,
+            MediaType::Type(media) => {
+                media[..prefix.len().min(media.len())].eq_ignore_ascii_case(prefix)
```
Would it be worth making `CaselessKey` transform the inner key into lowercase when instantiated? Then this comparison could read simpler (I think): `media.as_bytes().starts_with(prefix.as_bytes())`

igamigo commented on crates/rpc/src/server/accept.rs:
```diff
@@ -75,148 +182,232 @@ where
+    #[case::version_wildcard_minor("application/vnd.miden; version=1.*")]
+    #[case::version_wildcard_patch("application/vnd.miden; version=1.2.*")]
+    #[case::matching_network(
+        "application/vnd.miden; network=0x00000000000000000000000000000000000000000000000000000000deadbeef"
```
Should this be `genesis=`?


```suggestion
        "application/vnd.miden; genesis=0x00000000000000000000000000000000000000000000000000000000deadbeef"
```

I guess it works anyway because it's not validated in any way, but not sure if that was the intention.

igamigo reviewed (approved):
Overall LGTM

sergerad reviewed (approved):
LGTM

Mirko-von-Leipzig commented on crates/rpc/src/server/accept.rs:
```diff
@@ -75,148 +182,232 @@ where
+    #[case::invalid_version("application/vnd.miden; version=0x123")]
+    #[case::invalid_genesis("application/vnd.miden; genesis=aaa")]
+    #[case::version_too_old("application/vnd.miden; version=>1.2.3")]
+    #[case::version_too_new("application/vnd.miden; version=<1.2.3")]
```
What I was hoping this would add is the ability to exclude a buggy patch or version. But looking at [`VersionReq`](https://docs.rs/semver/latest/semver/struct.VersionReq.html) again, it uses comma's to separate definitions so I think this won't work at all :/

e.g. I was hoping to let client's skip buggy patch x.y.z by doing:

```
application/vnd.miden; version=>x.y.0, version<x.y.z,
application/vnd.miden; version=>x.y.z
```

This would let client's be more specific, but its a bit of a narrow use case so maybe I should revert to just `Version` since the `,` kinda screw it up.

Mirko-von-Leipzig replied:
Maybe this is a problem that gets solved by adding the server's RPC version to the content-type response header? Then the client can do as it wishes with that.

And the server can do normal semver version selection/rejection.

Mirko-von-Leipzig commented on crates/rpc/src/server/accept.rs:
```diff
@@ -1,61 +1,168 @@
+                    AcceptHeaderError::GenesisParsing(err, (*genesis).to_string())
+                })?;
+
+                if genesis != self.genesis_commitment {
```
I really want to. But technically its perfectly legal to combine media ranges like that.

I can't really think of a reason anyone might do so though, so maybe its okay to do so.

@igamigo thoughts?

Mirko-von-Leipzig commented:
Having looked into the specification some more, parsing gets really messy. I've since found a crate that actually adheres to the standard (or at least more so than the previous one's I'd discarded). I'll try replace my "parser" with this crate to judge if it might work.

For example, my parser doesn't handle quoted parameter values -- which notably would allow `,` inside the value. There are a bunch of these.

bobbinth commented:
> Having looked into the specification some more, parsing gets really messy. I've since found a crate that actually adheres to the standard (or at least more so than the previous one's I'd discarded). I'll try replace my "parser" with this crate to judge if it might work.
> 
> For example, my parser doesn't handle quoted parameter values -- which notably would allow `,` inside the value. There are a bunch of these.

We can also bring the scope of this PR back to the original issue to just add the `genesis` part to the `ACCEPT` header. For more comprehensive version compatibility checks we can create a separate issue and address it at a later point in time (IMO, it is a lower priority right now compared to checking if the genesis hashes match).

Mirko-von-Leipzig commented:
I've refactored this to use an external crate and revert the version requirements to what they were before. I've also updated the description to match the current reality.

If you could re-review please :)

Mirko-von-Leipzig commented on crates/rpc/src/server/accept.rs:
```diff
@@ -1,61 +1,186 @@
+
+impl AcceptHeaderLayer {
+    pub fn new(rpc_version: &Version, genesis_commitment: Word) -> Self {
+        let supported_versions = VersionReq {
+            comparators: vec![Comparator {
+                op: semver::Op::Exact,
+                major: rpc_version.major,
+                minor: rpc_version.minor.into(),
+                patch: None,
+                pre: semver::Prerelease::default(),
+            }],
+        };
```
I'm a bit on the fence about this; but I've left the current behavior as is.

The issue is that its a bit back-to-front. The client should be indicating which versions they're compatible with. For example, their RPC version might be patch compatible with our server's but they may require a new feature or addition that we introduced.

Its easy to do by swapping the client and server types around. The downside is that client's need to think a bit more about what they want.

bobbinth replied:
I think this is fine for now, but I would create an issue to revisit this in the future because in principle I agree: the client should be specifying the version that it is compatible with and if the node is within this specifications, the request should be accepted.

bobbinth reviewed (approved):
Looks good! Thank you! I left a couple of small comments inline.

Mirko-von-Leipzig commented on crates/rpc/src/server/accept.rs:
```diff
@@ -70,153 +195,195 @@ where
+impl FromStr for QValue {
+    type Err = QParsingError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
```
The format is a fixed point number with up-to 3 decimal places between zero and one, inclusive.

> (or maybe a fixed-point number)?

I assume you mean using some crate?

> If so, could we not parse it as a regular float

We should be able to, I initially dismissed this because I didn't think we could check the number of decimal places afterwards due to floating point idiosyncrasies. But I realize now we can check the length of the string instead :facepalm:

I'll try this out.

Mirko-von-Leipzig replied:
> Could we add a short doc comment explaining what the valid formats are?

I'll add it; the format is specified on the `QValue` type but I'll repeat it here.

Mirko-von-Leipzig replied:
I've done, its still a bit weird due to handling of decimal places. However my impl had a bug there so its definitely a positive :)

bobbinth reviewed (approved):
All looks good! Thank you!
